xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,1272 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
-	if err := edsB.waitForClose(ctx); err != nil {
-		t.Fatal(err)
+	// Verify that a successful RPC can be made.
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err = client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
+		t.Fatalf("EmptyCall() failed: %v", err)
 	}
 
-	// Make sure that the UpdateClientConnState() method on the CDS balancer
-	// returns error.
-	if err := cdsB.UpdateClientConnState(cdsCCS(clusterName, xdsC)); err != errBalancerClosed {
-		t.Fatalf("UpdateClientConnState() after close returned %v, want %v", err, errBalancerClosed)
+	// Retrieve the cds LB policy policy and close it.
+	var cdsBal balancer.Balancer
+	select {
+	case cdsBal = <-cdsBalancerCh:
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for cds LB policy to be created")
 	}
+	cdsBal.Close()
 
-	// Make sure that the ResolverErr() method on the CDS balancer does not
-	// forward the update to the EDS balancer.
-	rErr := errors.New("cdsBalancer resolver error")
-	cdsB.ResolverError(rErr)
-	sCtx, sCancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer sCancel()
-	if err := edsB.waitForResolverError(sCtx, rErr); err != context.DeadlineExceeded {
-		t.Fatal("ResolverError() forwarded to EDS balancer after Close()")
+	// Wait for the CDS resource to be not requested anymore.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for CDS resource to not requested")
+	case <-cdsResourceCanceledCh:
+	}
+	// Wait for the child policy to be closed.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for the child policy to be closed")
+	case <-childPolicyCloseCh:
 	}
 }
 
+// Tests the calling ExitIdle on the cds LB policy results in the call being
+// propagated to the child policy.
 func (s) TestExitIdle(t *testing.T) {
-	// This creates a CDS balancer, pushes a ClientConnState update with a fake
-	// xdsClient, and makes sure that the CDS balancer registers a watch on the
-	// provided xdsClient.
-	xdsC, cdsB, edsB, _, cancel := setupWithWatch(t)
-	defer func() {
-		cancel()
-		cdsB.Close()
-	}()
-
-	// Here we invoke the watch callback registered on the fake xdsClient. This
-	// will trigger the watch handler on the CDS balancer, which will attempt to
-	// create a new EDS balancer. The fake EDS balancer created above will be
-	// returned to the CDS balancer, because we have overridden the
-	// newChildBalancer function as part of test setup.
-	cdsUpdate := xdsresource.ClusterUpdate{
-		ClusterName: serviceName,
-		LBPolicy:    wrrLocalityLBConfigJSON,
-	}
-	wantCCS := edsCCS(serviceName, nil, false, wrrLocalityLBConfigJSON, noopODLBCfgJSON)
-	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer ctxCancel()
-	if err := invokeWatchCbAndWait(ctx, xdsC, cdsWatchInfo{cdsUpdate, nil}, wantCCS, edsB); err != nil {
+	// Register a wrapped cds LB policy for the duration of this test that
+	// allows explicitly calling ExitIdle on the LB policy.
```

Comment:
Same question here. Why does calling ExitIdle requiring wrapping?

Reply:
That's the only way to directly invoke it. Otherwise, you can call `cc.Connect()` and rely on it to call `ExitIdle()` on the top-level balancer which is `gsw` and rely on it to internally call `ExitIdle` on its child, which will be the CDS LB policy.

Reply:
Ohhhh I see. It's an e2e test (the whole point of this change), so you don't have a ref to the bal object in the system. I was thinking call Build{} on bb, but that's unit tests.